runtime/lua/vim/snippet.lua

```diff
@@ -0,0 +1,499 @@
+local G = require('vim.lsp._snippet_grammar')
+local snippet_group = 'vim/snippet'
+local snippet_ns = vim.api.nvim_create_namespace(snippet_group)
+
+--- Returns the 0-based cursor position.
+---
+--- @return integer, integer
+local function cursor_pos()
+  local cursor = vim.api.nvim_win_get_cursor(0)
+  return cursor[1] - 1, cursor[2]
+end
+
+--- Resolves variables (like `$name` or `${name:default}`) as follows:
+--- - When a variable is unknown (i.e.: its name is not recognized in any of the cases below), return `nil`.
+--- - When a variable isn't set, return its default (if any) or an empty string.
+---
+--- Note that in some cases, the default is ignored since it's not clear how to distinguish an empty
+--- value from an unset value (e.g.: `TM_CURRENT_LINE`).
+---
+--- @param var string
+--- @param default string
+--- @return string?
+local function resolve_variable(var, default)
+  --- @param str string
+  --- @return string
+  local function expand_or_default(str)
+    local expansion = vim.fn.expand(str) --[[@as string]]
+    return expansion == '' and default or expansion
+  end
+
+  if var == 'TM_SELECTED_TEXT' then
+    -- Snippets are expanded in insert mode only, so there's no selection.
+    return default
+  elseif var == 'TM_CURRENT_LINE' then
+    return vim.api.nvim_get_current_line()
+  elseif var == 'TM_CURRENT_WORD' then
+    return expand_or_default('<cword>')
+  elseif var == 'TM_LINE_INDEX' then
+    return tostring(vim.fn.line('.') - 1)
+  elseif var == 'TM_LINE_NUMBER' then
+    return tostring(vim.fn.line('.'))
+  elseif var == 'TM_FILENAME' then
+    return expand_or_default('%:t')
+  elseif var == 'TM_FILENAME_BASE' then
+    -- Not using '%:t:r' since we want to remove all extensions.
+    local filename_base = expand_or_default('%:t'):gsub('%.[^%.]*$', '')
+    return filename_base
+  elseif var == 'TM_DIRECTORY' then
+    return expand_or_default('%:p:h:t')
+  elseif var == 'TM_FILEPATH' then
+    return expand_or_default('%:p')
+  end
+
```

Comment:
I considered that but given that snippet expansion can only happen in visual mode, there won't be an active selection at that point. If we use marks then this would insert the last visual selection, and IMO most people are not expecting for that to be inserted in the snippet.

Reply:
> given that snippet expansion can only happen in visual mode

Why would that be the case?

Reply:
Can you think of a case where it wouldn't?

Reply:
you use completion and select an entry. You'd be in insert mode.

Reply:
ahh duh sorry, I brain farted. I meant to write insert mode all along.

Reply:
BTW I had brought this up in https://github.com/neovim/neovim/pull/25301#discussion_r1348178743.

Reply:
🤔 I also can't think about a case where you'd have a selection when expanding the snippet
And I agree that just using the last selection likely wouldn't be expected behavior. The only thing you perhaps could do was adding a check for visual mode and the using the marks.

Reply:
Actually I got a usecase.
This could be used e.g. for a "surround with html tag" function. You'd visually select something and then call the snippet function with `<$1:div>$TM_SELECTED_TEXT</$1>`. (not sure if the syntax is correct but just to give the idea)

Reply:
Please remember the goal here: expansion of LSP snippet completion results _and nothing else_.

Reply:
yes
this is about fully implementing the spec if possible (and without unexpected ux) which imo is doable with adding a check for visual mode. And additionaly to that I also gave an example where this might actually be useful. This wasn't intended as being a motivation for implementing it.

Reply:
Yes, but unless you can show a server that returns this _as a completion candidate_, it is not a relevant example in the context of this PR.

Reply:
👍 then this should be added to the "yet to be implemented in future prs" section imo since it's part of the spec, quite simple to implement afaict and actually has usecases with manual expansion

Reply:
No, because manual expansion is _not_ "yet to be implemented" in core. That will still be left to snippet plugins.

Reply:
Is there any good reason to not implement this thing apart from "there is no example where this is needed in completion items"? It would be quite small (check if mode is visual, get visual selection) and would allow to support the full lsp spec.
I don't see why you let snippet plugins rewrite their own expansion logic because of such a small thing just because it is more than you initially planned to support.

But in the end that's up to the core maintainers. I just write here from the perspective of a person that would perhaps use this functionality in a plugin. IMO the goal should be to support the full plugin spec if possible without it being too much effort and not just completion support.

Reply:
Please keep your comments to the scope of this PR. There will be a separate discussion about how to best expose this to (actual) snippet plugins; you're welcome to give your input then.

> But in the end that's up to the core maintainers.

Thank you for your understanding.

Reply:
Patience, grasshopper. Let @MariaSolOs finish this PR (and iron out any kinks found in testing) first.